pkg/apis/storage/types.go

```diff
@@ -80,3 +80,95 @@ type StorageClassList struct {
 	// Items is the list of StorageClasses
 	Items []StorageClass
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Captures the intent to attach or detach the specified volume to/from
+// the specified node.
+//
+// VolumeAttachment objects are non-namespaced.
+type VolumeAttachment struct {
+	metav1.TypeMeta
+
+	// Standard object metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// Specification of the desired attach/detach volume behavior.
+	// Populated by the Kubernetes system.
+	Spec VolumeAttachmentSpec
+
+	// Status of the VolumeAttachment request.
+	// Populated by the entity completing the attach or detach
+	// operation, i.e. the external-attacher.
+	Status VolumeAttachmentStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// VolumeAttachmentList is a collection of VolumeAttachment objects.
+type VolumeAttachmentList struct {
+	metav1.TypeMeta
+	// Standard list metadata
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+
+	// Items is the list of VolumeAttachments
+	Items []VolumeAttachment
+}
+
+// The specification of a VolumeAttachment request.
+type VolumeAttachmentSpec struct {
+	// The name of the external-attacher that MUST handle
+	// this request.
+	Attacher string
+
+	// The volume to attach.
+	Volume api.VolumeSource
```

Comment:
Is it intentional that VolumeAttachment is a cluster-scoped type that references `api.VolumeSource` instead of `api.PersistentVolumeSource`?

Reply:
Yes. It serves as memento of pod's VolumeSource in case the pod is deleted and we need to detach its volumes.

Reply:
there's no place to indicate the original namespace, pod name or pod uid... is none of that information needed to resolve the volume or referenced secrets?

Reply:
We moved to supporting PersistentVolumes only and we use namespaced SecretReferences there